internal/exec/stages/files/filesystemEntries.go

```diff
@@ -140,6 +142,42 @@ func (tmp linkEntry) node() types.Node {
 
 func (tmp linkEntry) create(l *log.Logger, u util.Util) error {
 	s := types.Link(tmp)
+	hard := s.Hard != nil && *s.Hard
+	st, err := os.Lstat(s.Path)
+	switch {
+	case os.IsNotExist(err):
+		break
+	case err != nil:
+		return fmt.Errorf("stat() failed on %s: %v", s.Path, err)
+	case hard:
+		// check that the file at that path points to the same inode as target
+		targetPath, err := u.JoinPath(s.Target)
+		if err != nil {
+			return fmt.Errorf("error resolving target path of hard link %s: %v", s.Path, err)
+		}
+		targetst, err := os.Lstat(targetPath)
+		if err != nil {
+			return fmt.Errorf("error creating hard link %s: target does not exist or stat() returned an err: %v", s.Path, err)
+		}
+		if !os.SameFile(st, targetst) {
+			return fmt.Errorf("error creating hard link %s: a file already exists at that path but is not the target and overwrite is false", s.Path)
+		}
+		l.Info("Hardlink %s to %s already exists, doing nothing", s.Path, s.Target)
+		return nil
+	case !hard:
+		// if the existing file is a symlink, check that its target is correct
+		if st.Mode()&os.ModeSymlink != 0 {
+			if target, err := os.Readlink(s.Path); err != nil {
+				return fmt.Errorf("error reading link at %s: %v", s.Path, err)
+			} else if filepath.Clean(target) != s.Target {
```

Comment:
Sanity-check: is `s.Target` also already cleaned? Or should this be `filepath.Clean(target) != filepath.Clean(s.Target)`?

Reply:
Thought it was, turns out it wasn't. Adding it here (and will PR with it to require clean paths for target).